Fix hard infinite loop on malformed stream management response

horazont opened the pull request:
If the server replies to a `<resume/>` request with a non-conformant `<failed/>` response, e.g. without a proper error condition, the `aioxmpp.node.Client` would get into a fun hard loop, until resource (memory) exhaustion eventually kills it off.

We fix this by discarding SM state if we get a stream error during the attempt of resuming SM. There may be edge cases where this is overzealous, but those could be fixed once they occur (I can't find any off the top of my head).

Fixes #382.

horazont committed:
Fix hard infinite loop on malformed stream management response

If the server replies to a `<resume/>` request with a non-conformant
`<failed/>` response, e.g. without a proper error condition, the
`aioxmpp.node.Client` would get into a fun hard loop, until resource
(memory) exhaustion eventually kills it off.

We fix this by discarding SM state if we get a stream error during the
attempt of resuming SM. There may be edge cases where this is
overzealous, but those could be fixed once they occur (I can't find
any off the top of my head).

Fixes #382.

jahrome commented:
Seems like it fixes the issue, no more endless loop and successful reconnection. Here is the output:

2022-10-10 04:03:19,104 [node] [ERROR]  stream failed: connection timeout (dead time hard limit exceeded)
2022-10-10 04:03:21,159 [protocol] [WARNING]  unhandled exception during parsing: ValueError('missing required member') (see full traceback later, this will kill the stream)
2022-10-10 04:03:21,161 [protocol] [ERROR]  unexpected exception while parsing stanza bubbled up through parser. stream so dead.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 583, in validate_contents
    obj = self.__get__(instance, type(instance))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 339, in __get__
    raise AttributeError(
AttributeError: attribute is unset (<aioxmpp.xso.model.Child object at 0x7fa9a2afd0> on instance of <class 'aioxmpp.nonza.SMFailed'>)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 491, in _rx_feed
    self._parser.feed(blob)
  File "/usr/lib/python3.9/xml/sax/expatreader.py", line 217, in feed
    self._parser.Parse(data, isFinal)
  File "../Modules/pyexpat.c", line 459, in EndElement
  File "/usr/lib/python3.9/xml/sax/expatreader.py", line 381, in end_element_ns
    self._cont_handler.endElementNS(pair, None)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1032, in endElementNS
    self._end_element_exception_handling()
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1018, in _end_element_exception_handling
    self.on_exception(exc)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 452, in _rx_exception
    raise exc from context
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1027, in endElementNS
    return self._driver.endElementNS(name, qname)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2543, in endElementNS
    self._send(("end",))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2527, in _send
    self._dest.send(value)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2678, in __call__
    cb((yield from cls.parse_events(ev_args, self._ctx)))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2103, in parse_events
    obj.validate()
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2368, in validate
    descriptor.validate_contents(self)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 585, in validate_contents
    raise ValueError("missing required member")
ValueError: missing required member
2022-10-10 04:03:21,251 [node] [ERROR]  failed to negotiate stream features
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 583, in validate_contents
    obj = self.__get__(instance, type(instance))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 339, in __get__
    raise AttributeError(
AttributeError: attribute is unset (<aioxmpp.xso.model.Child object at 0x7fa9a2afd0> on instance of <class 'aioxmpp.nonza.SMFailed'>)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 491, in _rx_feed
    self._parser.feed(blob)
  File "/usr/lib/python3.9/xml/sax/expatreader.py", line 217, in feed
    self._parser.Parse(data, isFinal)
  File "../Modules/pyexpat.c", line 459, in EndElement
  File "/usr/lib/python3.9/xml/sax/expatreader.py", line 381, in end_element_ns
    self._cont_handler.endElementNS(pair, None)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1032, in endElementNS
    self._end_element_exception_handling()
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1018, in _end_element_exception_handling
    self.on_exception(exc)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 452, in _rx_exception
    raise exc from context
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xml.py", line 1027, in endElementNS
    return self._driver.endElementNS(name, qname)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2543, in endElementNS
    self._send(("end",))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2527, in _send
    self._dest.send(value)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2678, in __call__
    cb((yield from cls.parse_events(ev_args, self._ctx)))
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2103, in parse_events
    obj.validate()
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 2368, in validate
    descriptor.validate_contents(self)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/xso/model.py", line 585, in validate_contents
    raise ValueError("missing required member")
ValueError: missing required member

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/node.py", line 962, in _main_impl
    features, sm_resumed = await self._negotiate_stream(
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/node.py", line 879, in _negotiate_stream
    resumed = await self._try_resume_stream_management(
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/node.py", line 847, in _try_resume_stream_management
    await self.stream.resume_sm(xmlstream)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/stream.py", line 2432, in resume_sm
    response = await protocol.send_and_wait_for(
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 916, in send_and_wait_for
    failure_future.result()
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 555, in data_received
    self._rx_feed(blob)
  File "/usr/local/lib/python3.9/dist-packages/aioxmpp/protocol.py", line 507, in _rx_feed
    raise errors.StreamError(
aioxmpp.errors.StreamError: stream error: {urn:ietf:params:xml:ns:xmpp-streams}internal-server-error ('Internal error while parsing XML. Client logs have more details.')
2022-10-10 04:03:21,287 [node] [WARNING]  discarding SM state because of error during negotiation

horazont commented:
Excellent, thanks!

I'll then see what's up with the CI and make sure to release a bugfix release soon-ish.

horazont committed:
Update test matrix in CI

- 3.5 and 3.6 are no longer supported, and we need to test 3.11.
- Prosody 0.11 vanished from repositories, updated to 0.12
- legacy ejabberd fails in mysterious ways -> removing
- current ejabberd doesn't start right away and I'm tired

horazont committed:
tests: update regexes for Python 3.11
